File: backend/routes/grades.py
from flask import Blueprint, request, jsonify
from models.grade import Grade
from models.submission import Submission
from models.week import Week
from middleware.auth import token_required, admin_required
from extensions import db

grades_bp = Blueprint("grades", __name__)


@grades_bp.route("/", methods=["POST"])
@admin_required
def create_or_update_grade(current_user):
    data = request.get_json()
    if not data:
        return jsonify({"error": "No data provided"}), 400

    submission_id = data.get("submission_id")
    if not submission_id:
        return jsonify({"error": "Submission ID is required"}), 400

    submission = Submission.query.filter_by(id=submission_id).first()
    if not submission:
        return jsonify({"error": "Submission not found"}), 404

    try:
        grade = Grade.query.filter_by(submission_id=submission_id).first()
        if not grade:
            grade = Grade(submission_id=submission_id)
            db.session.add(grade)

        grade.ui = data.get("ui", grade.ui)
        grade.functionality = data.get("functionality", grade.functionality)
        grade.github = data.get("github", grade.github)
        grade.documentation = data.get("documentation", grade.documentation)
        grade.innovation = data.get("innovation", grade.innovation)
        grade.weekly_progress = data.get("weekly_progress", grade.weekly_progress)
        grade.published = data.get("published", grade.published)

        grade.calculate_total()
        db.session.commit()

        result = {
            **grade.to_dict(),
            "week": submission.week.to_dict() if submission and submission.week else None,
            "user": submission.user.to_dict() if submission and submission.user else None,
        }

        return jsonify({"message": "Grade saved", "grade": result}), 200
    except Exception as e:
        db.session.rollback()
        return jsonify({"error": f"Failed to save grade: {str(e)}"}), 500


@grades_bp.route("/submission/<submission_id>", methods=["GET"])
@token_required
def get_grade(current_user, submission_id):
    grade = Grade.query.filter_by(submission_id=submission_id).first()
    if not grade:
        return jsonify({"error": "Grade not found"}), 404
    if current_user.role != "admin" and not grade.published:
        return jsonify({"error": "Grade not published yet"}), 403

    submission = Submission.query.get(submission_id)
    result = {
        **grade.to_dict(),
        "week": submission.week.to_dict() if submission and submission.week else None,
        "user": submission.user.to_dict() if submission and submission.user else None,
    }

    return jsonify({"grade": result}), 200


@grades_bp.route("/publish/<submission_id>", methods=["POST"])
@admin_required
def publish_grade(current_user, submission_id):
    grade = Grade.query.filter_by(submission_id=submission_id).first()
    if not grade:
        return jsonify({"error": "Grade not found"}), 404
    try:
        grade.published = True
        db.session.commit()

        submission = Submission.query.get(submission_id)
        result = {
            **grade.to_dict(),
            "week": submission.week.to_dict() if submission and submission.week else None,
            "user": submission.user.to_dict() if submission and submission.user else None,
        }

        return jsonify({"message": "Grade published", "grade": result}), 200
    except Exception as e:
        db.session.rollback()
        return jsonify({"error": f"Failed to publish grade: {str(e)}"}), 500


@grades_bp.route("/candidate/<candidate_id>", methods=["GET"])
@token_required
def get_candidate_grades(current_user, candidate_id):
    if current_user.role != "admin" and current_user.id != candidate_id:
        return jsonify({"error": "Access denied"}), 403

    project_id = request.args.get("project_id")
    query = (
        db.session.query(Grade)
        .join(Submission, Grade.submission_id == Submission.id)
        .join(Week, Submission.week_id == Week.id)
        .filter(Submission.user_id == candidate_id)
    )
    if project_id:
        query = query.filter(Week.project_id == project_id)

    grades = query.all()

    result = []
    for g in grades:
        if current_user.role == "admin" or g.published:
            sub = Submission.query.get(g.submission_id)
            result.append({
                **g.to_dict(),
                "week": sub.week.to_dict() if sub and sub.week else None,
            })

    return jsonify({"grades": result}), 200


@grades_bp.route("/all", methods=["GET"])
@admin_required
def get_all_grades(current_user):
    grades = Grade.query.all()
    result = []
    for g in grades:
        sub = Submission.query.get(g.submission_id)
        result.append({
            **g.to_dict(),
            "week": sub.week.to_dict() if sub and sub.week else None,
            "user": sub.user.to_dict() if sub and sub.user else None,
        })
    return jsonify({"grades": result}), 200


@grades_bp.route("/stats/candidate/<candidate_id>", methods=["GET"])
@token_required
def get_candidate_stats(current_user, candidate_id):
    if current_user.role != "admin" and current_user.id != candidate_id:
        return jsonify({"error": "Access denied"}), 403

    project_id = request.args.get("project_id")
    query = (
        db.session.query(Grade)
        .join(Submission, Grade.submission_id == Submission.id)
        .join(Week, Submission.week_id == Week.id)
        .filter(Submission.user_id == candidate_id, Grade.published == True)
    )
    if project_id:
        query = query.filter(Week.project_id == project_id)

    grades = query.all()

    weekly_scores = []
    for g in grades:
        sub = Submission.query.get(g.submission_id)
        weekly_scores.append({
            "week": sub.week.week_title if sub and sub.week else "Unknown",
            "total": g.total,
            "grade": g.grade,
        })

    total_score = sum(g.total for g in grades) if grades else 0
    avg_score = total_score / len(grades) if grades else 0

    return jsonify({
        "weekly_scores": weekly_scores,
        "average_score": round(avg_score, 2),
        "total_graded": len(grades),
    }), 200


@grades_bp.route("/stats/overview", methods=["GET"])
@admin_required
def get_overview_stats(current_user):
    from models.user import User

    total_candidates = User.query.filter_by(role="candidate").count()
    total_submissions = Submission.query.count()
    published_grades = Grade.query.filter_by(published=True).all()
    avg_grade = (
        sum(g.total for g in published_grades) / len(published_grades)
        if published_grades
        else 0
    )
    pending_reviews = Submission.query.filter(
        ~Submission.id.in_(db.session.query(Grade.submission_id))
    ).count()

    return jsonify({
        "total_candidates": total_candidates,
        "total_submissions": total_submissions,
        "pending_reviews": pending_reviews,
        "average_grade": round(avg_grade, 2),
    }), 200

File: frontend/lib/services/grades.ts
import { supabase } from "../supabase/client";
import type { Grade } from "@/types";

function calculateGradeLetter(total: number): string {
  if (total >= 90) return "A+";
  if (total >= 80) return "A";
  if (total >= 70) return "B";
  if (total >= 60) return "C";
  return "Needs Improvement";
}

export const gradesService = {
  async createOrUpdate(data: Record<string, unknown>): Promise<{ message: string; grade: Grade }> {
    const submissionId = data.submission_id as string;
    const ui = Number(data.ui || 0);
    const functionality = Number(data.functionality || 0);
    const github = Number(data.github || 0);
    const documentation = Number(data.documentation || 0);
    const innovation = Number(data.innovation || 0);
    const weeklyProgress = Number(data.weekly_progress || 0);
    const published = Boolean(data.published ?? false);

    const total = ui + functionality + github + documentation + innovation + weeklyProgress;
    const gradeLetter = calculateGradeLetter(total);

    const { data: existing } = await supabase
      .from("grades")
      .select("id")
      .eq("submission_id", submissionId)
      .single();

    let gradeObj;
    if (existing) {
      const { data: updated, error } = await supabase
        .from("grades")
        .update({
          ui,
          functionality,
          github,
          documentation,
          innovation,
          weekly_progress: weeklyProgress,
          total,
          grade: gradeLetter,
          published,
        })
        .eq("id", existing.id)
        .select("*, submission:submissions(*, week:weeks(*), user:profiles(*))")
        .single();
      if (error) throw error;
      gradeObj = updated;
    } else {
      const { data: inserted, error } = await supabase
        .from("grades")
        .insert({
          submission_id: submissionId,
          ui,
          functionality,
          github,
          documentation,
          innovation,
          weekly_progress: weeklyProgress,
          total,
          grade: gradeLetter,
          published,
        })
        .select("*, submission:submissions(*, week:weeks(*), user:profiles(*))")
        .single();
      if (error) throw error;
      gradeObj = inserted;
    }

    return { message: "Grade saved", grade: gradeObj as unknown as Grade };
  },

  async getBySubmission(submissionId: string): Promise<{ grade: Grade }> {
    const { data, error } = await supabase
      .from("grades")
      .select("*, submission:submissions(*, week:weeks(*), user:profiles(*))")
      .eq("submission_id", submissionId)
      .single();

    if (error || !data) throw new Error("Grade not found");
    return { grade: data as unknown as Grade };
  },

  async publish(submissionId: string): Promise<{ message: string; grade: Grade }> {
    const { data, error } = await supabase
      .from("grades")
      .update({ published: true })
      .eq("submission_id", submissionId)
      .select("*, submission:submissions(*, week:weeks(*), user:profiles(*))")
      .single();

    if (error || !data) throw new Error("Failed to publish grade");
    return { message: "Grade published", grade: data as unknown as Grade };
  },

  async getCandidateGrades(candidateId: string, projectId?: string): Promise<{ grades: Grade[] }> {
    let query = supabase
      .from("grades")
      .select("*, submission:submissions!inner(*, week:weeks!inner(*))")
      .eq("submission.user_id", candidateId);

    if (projectId) {
      query = query.eq("submission.week.project_id", projectId);
    }

    const { data, error } = await query;
    if (error) throw error;
    return { grades: (data || []) as unknown as Grade[] };
  },

  async getAll(): Promise<{ grades: Grade[] }> {
    const { data, error } = await supabase
      .from("grades")
      .select("*, submission:submissions(*, week:weeks(*), user:profiles(*))");

    if (error) throw error;
    return { grades: (data || []) as unknown as Grade[] };
  },

  async getCandidateStats(candidateId: string, projectId?: string) {
    const { grades } = await this.getCandidateGrades(candidateId, projectId);
    const published = grades.filter((g) => g.published);

    const weeklyScores = published.map((g) => ({
      week: g.submission_id,
      total: g.total,
      grade: g.grade,
    }));

    const totalScore = published.reduce((acc, curr) => acc + curr.total, 0);
    const avgScore = published.length ? totalScore / published.length : 0;

    return {
      weekly_scores: weeklyScores,
      average_score: Math.round(avgScore * 100) / 100,
      total_graded: published.length,
    };
  },

  async getOverviewStats() {
    const { data: candidates } = await supabase.from("profiles").select("id").eq("role", "candidate");
    const { data: subs } = await supabase.from("submissions").select("id");
    const { data: grades } = await supabase.from("grades").select("total").eq("published", true);

    const avg = grades && grades.length ? grades.reduce((a, b) => a + b.total, 0) / grades.length : 0;

    return {
      total_candidates: candidates?.length || 0,
      total_submissions: subs?.length || 0,
      pending_reviews: Math.max(0, (subs?.length || 0) - (grades?.length || 0)),
      average_grade: Math.round(avg * 100) / 100,
    };
  },
};
